Fix: clearing characters in search bar

belljustin opened the pull request:
Refs #64

When the searchbar was changed it was sending a debounced dispatch that
executed a search request. EXCEPT when the bar was empty, it sent a
dispatch to clear the search state. However after the debounce timeout, the
search request action would still dispatch.

Now the clear dispatch is within the debounced method such that it
replaces the previous func

belljustin committed:
Fix: clearing characters in search bar
Refs #64
When the searchbar was changed it was sending a debounced dispatch that
executed a search request. EXCEPT when the bar was empty, it sent a
dispatch to clear the search state. However after the debounce timeout, the
search request action would still dispatch.

Now the clear dispatch is within in the debounced method such that it
replaces the previous func

wabain commented:
I think the race condition here is actually more general. The issue is that a SEARCH_REQUEST action with a status of success or error will always overwrite the current search. With the clear action not being debounced, a clear is the most obvious way to get the race, but you could get the same problem with a series of actions like the following:

```js
{type: SEARCH_REQUEST, query: 'A', status: PENDING}
{type: SEARCH_REQUEST, query: 'B', status: PENDING}
{type: SEARCH_REQUEST, query: 'B', status: SUCCESS, ...}
{type: SEARCH_REQUEST, query: 'A', status: SUCCESS, ...}
```

belljustin committed:
WIP: addresses race condition with search request
Refs #64

When an asynchronous call is made for a new search request, the current
query state is recorded. When the response is recieved, we check if the
query state is still the same. If it is, we dispatch the succesful
action

wabain commented on misirlou/frontend/js/action-creators/search.js:
```diff
@@ -92,6 +96,22 @@ const execSearch = debounce((query, dispatch, getSuggestions) =>
+        Search.get(query).then(
+            response =>
+            {
+                if (start_state.query === getState().search.current.query)
```
This check should occur in the error case as well. One way to do this would be to return the `getSearchAction` output from both promises, then to add a chained `.then` which does the check and the final dispatch.

wabain commented:
Apart from the bit I commented on the logic here looks good to me. There are a few JavaScript issues, mostly around variable declaration. Running `npm lint-js` will highlight those. (It looks like there are unrelated errors from page.js, cc @jeromepl. I guess the linting would benefit from more visibility.)

belljustin committed:
Fix: catches the onRejected case as well

belljustin commented:
Should be ready for merge now, @lexpar
